README
Filebrowser plugin for the DeaDBeeF audio player

Available on SF.net and GitLab:
    http://sourceforge.net/projects/deadbeef-fb/
    https://gitlab.com/zykure/deadbeef-fb

Binary and source releases for download:
    https://gitlab.com/zykure/deadbeef-fb/tree/release
    https://sourceforge.net/projects/deadbeef-fb/files/ (outdated)

Please report bugs or post feature requests on GitLab:
    https://gitlab.com/zykure/deadbeef-fb/issues


Copyright (C) 2011-2016 Jan D. Behrens <zykure@web.de>

Based on Geany treebrowser plugin:
    treebrowser.c - v0.20
    Copyright 2010 Adrian Dimitrov <dimitrov.adrian@gmail.com>


Check out the player itself here:
    http://deadbeef.sourceforge.net/


==========================================================

1. Compiling etc.

* You can also grab the binary archive with the precompiled plugin
  if you don't want to compile for yourself. See below how to install
  the plugin when you already have the binary files.

* To build the plugin from source, you'll need the deadbeef and
  GTK2/GTK3 headers (deadbeef-dev, libgtk2.0-dev, libgtk3.0-dev).

* To get the sources from git, use the following commands:
    First donwload:   git clone git://deadbeef-fb.git.sourceforge.net/gitroot/deadbeef-fb/deadbeef-fb
    Update the tree:  git pull

* Then just follow the instructions in the INSTALL file.
  Basically, you need to run
    ./autogen.sh
    ./configure
    make
  to compile the sources. After that, the plugin is ready for install.

* By default, both GTK2 and GTK3 versions of the plugin are built. You
  can disable either one by passing the --disable-gtk2|3 flag to
  configure:
    ./configure --disable-gtk3
  Then continue as described above.


==========================================================

2. Install

* You can either use one of the provided install scripts, or use
  "make install" as mentioned in the INSTALL file.

* To quickly install the plugin into your home directory, use the
  "userinstall.sh" script. You don't need to be root to do that!
  It will put the plugin file into $HOME/.local/lib/deadbeef/.

* To quickly install the plugin into your system directory, use the
  "quickinstall.sh" script.
  It will put the plugin file into /usr/local/lib/deadbeef/.

* If you want to install the plugin into your system directory (e.g.
  /usr), it's better to compile the program and use "make install", as
  this also takes care of other files like translations! By default,
  the plugin will be installed to /usr/local/ using this method.


==========================================================

3. Problems?

* If DeaDBeeF stops working after you've installed the plugin, just
  remove the .so file from the directory you choosed above.


ChangeLog
commit beaf1a001f6adc64ba2efb093fee73192f8652a1
Author: duzda <duzda666@gmail.com>
Date:   Fri Apr 8 18:23:04 2022 +0200

    typos

commit 154c63086a1fb62d4508353a1ba89ac439d385c4
Author: duzda <duzda666@gmail.com>
Date:   Fri Apr 8 18:08:55 2022 +0200

    bigger minimum expand character limit

commit 4510a330590fb41db5222aabf0815cc169506624
Author: duzda <duzda666@gmail.com>
Date:   Fri Apr 8 18:05:41 2022 +0200

    added collapse all button

commit 4c0aaa2db3393babd91bd3973176c73d88da0673
Author: duzda <duzda666@gmail.com>
Date:   Fri Apr 8 17:47:45 2022 +0200

    fix: compile error

commit 355e6142606270e7ae0fd68d8fcdb3f1ada295ca
Merge: 47bc3fe ca06166
Author: zykure <zykure@protonmail.com>
Date:   Mon May 7 17:10:10 2018 +0000

    Merge branch 'master' into 'master'
    
    Fix corrupted coverarts crashing deadbeef (#19)
    
    See merge request zykure/deadbeef-fb!6

commit ca061662b68cb502dc8e863b1160b7ea92ee4919
Author: Dante <dsdante@protonmail.com>
Date:   Sat Apr 28 19:15:37 2018 +0300

    Fix corrupted coverarts crashing deadbeef (#19)

commit 47bc3fe0fe151dfbb75f90f9e4d5e9a7affd280a
Author: Jan Behrens <zykure42@gmail.com>
Date:   Tue Feb 14 17:32:20 2017 +0100

    gtk_menu_popup_at_pointer is availabe in gtk>=3.22

commit 9ad84e1308d8429111760ff31802982aea504657
Merge: 1562809 fd881f3
Author: zykure <zykure42@gmail.com>
Date:   Tue Feb 14 16:22:23 2017 +0000

    Merge branch 'master' into 'master'
    
    gtk_menu_popup deprecated
    
    See merge request !5

commit fd881f3e8a30f59a2636817e8314468aa0ea0203
Author: Guillaume Castagnino <casta@xwing.info>
Date:   Sat Dec 31 10:24:00 2016 +0100

    gtk_menu_popup deprecated
    
    As of gtk-3.22, gtk_menu_popup is deprecated and compilation with it
    generates an error.
    The documentation provides replacement functions:
    https://developer.gnome.org/gtk3/unstable/GtkMenu.html#gtk-menu-popup
    
    Here is my proposal, that seems to do the job.

commit 156280934b790bf260f235522140013a64f9b9fd
Author: zykure <zykure42@gmail.com>
Date:   Fri Aug 26 06:51:16 2016 +0000

    Update README

commit ee4f6b3e280c7e985b04d463decc0545387ae7db
Author: Jan Behrens <zykure@web.de>
Date:   Tue Feb 16 09:20:51 2016 +0100

    Add option to disable coverart icons

commit 6e0a0aa8d2b56f608581ff0bc6fcd594d3f7669b
Author: Jan Behrens <zykure@web.de>
Date:   Tue Feb 16 09:11:17 2016 +0100

    Ensure that icons are larger at least 24px, see #16

commit 7785ff9c3f963a8dc54926485c0ce96ccdf81650
Author: Jan Behrens <zykure@web.de>
Date:   Fri Feb 12 18:57:05 2016 +0100

    Fix GTK2 colors

commit 971fa7567eb46fd3fa958bfd9a16541197e64fd7
Author: Jan Behrens <zykure@web.de>
Date:   Tue Feb 9 23:01:04 2016 +0100

    Fix GTK3 version checks (gtk_label_set_xalign)

commit d62b1f12e3aad34487d73dd5dc726fe5976fed74
Author: Jan Behrens <zykure@web.de>
Date:   Mon Feb 8 19:02:21 2016 +0100

    Some more tweaking

commit c88d415a78aa019b9518c9fb3df5416bfd66f5ec
Author: Jan Behrens <zykure@web.de>
Date:   Mon Feb 8 14:15:11 2016 +0100

    Finally fix GTK3 colors (must override bg-image and borders as well... duh!)

commit d60f09fcb3770c6a454f9b4ce6917c69387f0171
Author: Jan Behrens <zykure@web.de>
Date:   Tue Jan 19 23:45:35 2016 +0100

    Improve keybindings, make ctrl work with double-click (see !4)

commit e4e497993624b3469af917957c7f8625740488a7
Author: Jan Behrens <zykure@web.de>
Date:   Tue Jan 19 10:47:14 2016 +0100

    Add rename option (GTK3 only), fixes #4

commit 910cddea8284a60053ff6d20082daa261e4ba554
Author: Jan Behrens <zykure@web.de>
Date:   Mon Jan 18 23:37:55 2016 +0100

    version bump

commit ee572932cbca1a971e51c08b4b9e4e552e82e990
Merge: 39192b3 0e65da7
Author: Jan Behrens <zykure@web.de>
Date:   Mon Jan 18 23:37:30 2016 +0100

    Merge branch 'ximm/deadbeef-fb-feature-advanced-controls'

commit 0e65da74e990ec1d459817e80d4c8328a98c9c02
Author: Jan Behrens <zykure@web.de>
Date:   Mon Jan 18 23:30:06 2016 +0100

    Rework some of the new code, change some keybindings, update help text

commit d26a6f8ca800cbb097a3e02c4c04235fe756b8b9
Author: Jan Behrens <zykure@web.de>
Date:   Mon Jan 18 22:34:57 2016 +0100

    fix errors/warnings

commit b8b37792d00f4d0147477a66b57ebd8f9eac9308
Author: Tobias Bengfort <tobias.bengfort@gmx.net>
Date:   Sun Jan 17 20:39:17 2016 +0100

    add keyboard controls

commit d1b14f13150d3632cf3d00cfada5d0a4a70016a3
Author: Tobias Bengfort <tobias.bengfort@gmx.net>
Date:   Sun Jan 17 21:58:06 2016 +0100

    grab focus on click

commit 3d7654aef97a2054c133188e1e4838998616fc9e
Author: Tobias Bengfort <tobias.bengfort@gmx.net>
Date:   Sun Jan 17 20:38:54 2016 +0100

    activate on double click

commit 70798bd82db91ec283061f3b37c98b3248f89a63
Author: Tobias Bengfort <tobias.bengfort@gmx.net>
Date:   Sun Jan 17 20:32:26 2016 +0100

    Fix keyboard navigation

commit 39192b39dfbb6bae0b483a614d5b6d8af3871479
Author: Jan Behrens <zykure@web.de>
Date:   Wed Jan 6 21:05:08 2016 +0100

    Fix those damn build scripts again

commit c2afbbe5a32d1fa997b20b962eacb012bc1b93a2
Author: Jan Behrens <zykure@web.de>
Date:   Wed Jan 6 20:40:58 2016 +0100

    Fix build scripts; bump version

commit f3a904e91bace65f35ba87d849c1cd00ec7f15aa
Merge: 4c56899 e652321
Author: zykure <zykure42@gmail.com>
Date:   Mon Jan 4 08:29:36 2016 +0000

    Merge branch 'fixes' into 'master'
    
    Fixes
    
    Don't do debug printfs in committed code.
    
    Also fix the argument order for gtk_drag_check_threshold so it actually works.
    
    
    See merge request !3

commit e6523218e81bee33f5407dd2fd1823ead39d2775
Author: Nicolai Syvertsen <saivert@saivert.com>
Date:   Tue Dec 29 05:11:18 2015 +0100

    Fix gtk_drag_check_threshold argument order

commit aafab03ea653daf5a14498a8d5a30da1cc444c5b
Author: Nicolai Syvertsen <saivert@saivert.com>
Date:   Tue Dec 29 05:08:07 2015 +0100

    Remove debug printf

commit 4c5689941d7ac170b8f0969c7c776c26f9c3e3df
Author: Jan Behrens <zykure@web.de>

utils.c
/* UTILITY FUNCTIONS */

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <gtk/gtk.h>
#include "utils.h"


/* Get pixbuf icon from current icon theme */
GdkPixbuf *
utils_pixbuf_from_stock (const gchar *icon_name, gint size)
{
    GtkIconTheme *icon_theme = gtk_icon_theme_get_default();
    if (icon_theme)
        return gtk_icon_theme_load_icon (icon_theme, icon_name, size, 0, NULL);

    return NULL;
}

/* Check if two strings are exactly the same */
gboolean
utils_str_equal (const gchar *a, const gchar *b)
{
    /* (taken from libexo from os-cillation) */
    if (a == NULL && b == NULL)
        return TRUE;
    else if (a == NULL || b == NULL)
        return FALSE;

    while (*a == *b++)
        if (*a++ == '\0')
            return TRUE;

    return FALSE;
}

/* Compare two strings, ignoring case differences */
gint
utils_str_casecmp (const gchar *s1, const gchar *s2)
{
    gchar *tmp1, *tmp2;
    gint result;

    g_return_val_if_fail (s1 != NULL, 1);
    g_return_val_if_fail (s2 != NULL, -1);

    tmp1 = g_strdup (s1);
    tmp2 = g_strdup (s2);

    /* first ensure strings are UTF-8 */
    if (! g_utf8_validate (s1, -1, NULL))
        setptr (tmp1, g_locale_to_utf8 (s1, -1, NULL, NULL, NULL));
    if (! g_utf8_validate (s2, -1, NULL))
        setptr (tmp2, g_locale_to_utf8 (s2, -1, NULL, NULL, NULL));

    if (tmp1 == NULL) {
        g_free (tmp2);
        return 1;
    }
    if (tmp2 == NULL) {
        g_free (tmp1);
        return -1;
    }

    /* then convert the strings into a case-insensitive form */
    setptr (tmp1, g_utf8_strdown (tmp1, -1));
    setptr (tmp2, g_utf8_strdown (tmp2, -1));

    /* compare */
    result = strcmp (tmp1, tmp2);
    g_free (tmp1);
    g_free (tmp2);
    return result;
}

/* Get list of filenames inside a directory */
GSList *
utils_get_file_list_full (const gchar *path, gboolean full_path, gboolean sort, GError **error)
{
    GSList *list = NULL;
    GDir *dir;
    const gchar *filename;

    if (error)
        *error = NULL;
    g_return_val_if_fail (path != NULL, NULL);

    dir = g_dir_open (path, 0, error);
    if (dir == NULL)
        return NULL;

    foreach_dir (filename, dir) {
        list = g_slist_prepend (list, full_path ? g_build_path (G_DIR_SEPARATOR_S, path, filename, NULL) : g_strdup (filename));
    }
    g_dir_close (dir);

    /* sorting last is quicker than on insertion */
    if (sort)
        list = g_slist_sort (list, (GCompareFunc) utils_str_casecmp);
    else
        list = g_slist_reverse (list);
    return list;
}

GSList *
utils_get_file_list (const gchar *path, guint *length, gboolean sort, GError **error)
{
    GSList *list = utils_get_file_list_full (path, FALSE, sort, error);

    if (length)
        *length = g_slist_length (list);
    return list;
}


/* Convert text in local encoding to UTF8 */
gchar *
utils_get_utf8_from_locale(const gchar *locale_text)
{
    gchar *utf8_text;

    if (! locale_text)
        return NULL;
    utf8_text = g_locale_to_utf8 (locale_text, -1, NULL, NULL, NULL);
    if (utf8_text == NULL)
        utf8_text = g_strdup (locale_text);
    return utf8_text;
}

/* Get current home directory */
gchar *
utils_expand_home_dir (const gchar *path)
{
    gchar *expanded_path = g_strdup (path);

    if (g_strrstr (expanded_path, "~"))
    {
        gchar **split = g_strsplit (expanded_path, "~", 0);
        g_free (expanded_path);
        expanded_path = g_strjoinv (g_getenv ("HOME"), split);
        g_strfreev (split);
    }

    if (g_strrstr (expanded_path, "$HOME"))
    {
        gchar **split = g_strsplit (expanded_path, "$HOME", 0);
        g_free (expanded_path);
        expanded_path = g_strjoinv (g_getenv ("HOME"), split);
        g_strfreev (split);
    }

    if (g_strrstr (expanded_path, "${HOME}"))
    {
        gchar **split = g_strsplit (expanded_path, "${HOME}", 0);
        g_free (expanded_path);
        expanded_path = g_strjoinv (g_getenv ("HOME"), split);
        g_strfreev (split);
    }

    return expanded_path;
}

/* Get URI tooltip from URI, escaping ampersands ('&') */
gchar *
utils_tooltip_from_uri (const gchar *uri)
{
    if (! uri)
        return NULL;
    gchar **strings = g_strsplit (uri, "&", 0);
    gchar *result = g_strjoinv ("&amp;", strings);
    g_strfreev (strings);
    return result;
}

/* Get cache path for directory icon of URI */
gchar *
utils_make_cache_path (const gchar *uri, gint imgsize, gboolean scale)
{
    /* Path for cached icon is constructed like this:
     *     $XDG_CACHE_HOME/deadbeef-fb/icons/<imgsize>/<uri-without-separators>.png
     * If $XDG_CACHE_HOME is undefined, $HOME/.cache/deadbeef-fb/ is used instead.
     *
     * Example: The coverart image
     *     /home/user/Music/SomeArtist/Album01/cover.jpg
     * will lead to the cached at
     *     /home/user/.cache/deadbeef-fb/24/home_user_Music_SomeArtist_Album01.png
     */
    const gchar *cache = g_getenv ("XDG_CACHE_HOME");
    GString *path, *fullpath;
    gchar *cachedir, *fname;

    path = g_string_sized_new (256);  // reasonable initial size
    g_string_printf (path,
                    cache ? "%s/deadbeef-fb/icons/%d/%s/" : "%s/.cache/deadbeef-fb/icons/%d/%s/",
                    cache ? cache : g_getenv ("HOME"),
                    imgsize,
                    scale ? "scaled" : "");
    cachedir = g_string_free (path, FALSE);

    /* Create path if it doesn't exist already */
    if (! g_file_test (cachedir, G_FILE_TEST_IS_DIR))
        utils_check_dir (cachedir, 0755);

    fullpath = g_string_new (g_strdup (cachedir));

    fname = g_strdup (uri);
    for (gchar *p = fname+1; *p; p++) {
        if ((*p == G_DIR_SEPARATOR_S[0]) || (*p == ' ')) {
            *p = '_';
        }
    }

    g_string_append_printf (fullpath, "/%s.png", fname);

    g_free (cachedir);
    g_free (fname);

    return g_string_free (fullpath, FALSE);
}

/* Copied from  <deadbeef>/plugins/artwork/artwork.c  with few adjustments */
gint
utils_check_dir (const gchar *dir, mode_t mode)
{
    gchar *tmp = g_strdup (dir);
    gchar *slash = tmp;
    struct stat stat_buf;
    do
    {
        slash = strstr (slash+1, "/");
        if (slash)
            *slash = 0;
        if (-1 == stat (tmp, &stat_buf)) {
            int err_no = mkdir (tmp, mode);
            if (0 != err_no) {
                fprintf (stderr, "Failed to create %s (%d)\n", tmp, err_no);
                g_free (tmp);
                return 0;
            }
        }
        if (slash)
            *slash = '/';
    }
    while (slash);

    g_free (tmp);
    return 1;
}

void
utils_construct_style (GtkWidget *widget, const gchar *bgcolor, const gchar *fgcolor, const gchar *bgcolor_sel, const gchar *fgcolor_sel)
{
    if (! widget)
        return;

    GString *style = g_string_new ("");
#if !GTK_CHECK_VERSION(3,0,0)
    style = g_string_append (style, "style \"deadbeef-filebrowser\" { \n");
    if (strlen(bgcolor) > 0)       g_string_append_printf (style, "    GtkTreeView::even-row-color = \"%s\" \n", bgcolor);
    if (strlen(bgcolor) > 0)       g_string_append_printf (style, "    GtkTreeView::odd-row-color = \"%s\" \n", bgcolor);
    if (strlen(bgcolor) > 0)       g_string_append_printf (style, "    base[NORMAL]   = \"%s\" \n", bgcolor);
    if (strlen(bgcolor_sel) > 0)   g_string_append_printf (style, "    base[SELECTED] = \"%s\" \n", bgcolor_sel);
    if (strlen(bgcolor_sel) > 0)   g_string_append_printf (style, "    base[ACTIVE]   = \"%s\" \n", bgcolor_sel);
    if (strlen(fgcolor) > 0)       g_string_append_printf (style, "    text[NORMAL]   = \"%s\" \n", fgcolor);
    if (strlen(fgcolor_sel) > 0)   g_string_append_printf (style, "    text[SELECTED] = \"%s\" \n", fgcolor_sel);
    if (strlen(fgcolor_sel) > 0)   g_string_append_printf (style, "    text[ACTIVE]   = \"%s\" \n", fgcolor_sel);
    if (strlen(bgcolor) > 0)       g_string_append_printf (style, "    bg[NORMAL]     = \"%s\" \n", bgcolor);
    if (strlen(bgcolor_sel) > 0)   g_string_append_printf (style, "    bg[SELECTED]   = \"%s\" \n", bgcolor_sel);
    if (strlen(bgcolor_sel) > 0)   g_string_append_printf (style, "    bg[ACTIVE]     = \"%s\" \n", bgcolor_sel);
    if (strlen(fgcolor) > 0)       g_string_append_printf (style, "    fg[NORMAL]     = \"%s\" \n", fgcolor);
    if (strlen(fgcolor_sel) > 0)   g_string_append_printf (style, "    fg[SELECTED]   = \"%s\" \n", fgcolor_sel);
    if (strlen(fgcolor_sel) > 0)   g_string_append_printf (style, "    fg[ACTIVE]     = \"%s\" \n", fgcolor_sel);
    style = g_string_append (style, "} \n");
    style = g_string_append (style, "widget \"*.deadbeef_filebrowser_treeview\" style \"deadbeef-filebrowser\" \n");
#else
    style = g_string_append (style, "GtkTreeView { \n");
    style = g_string_append (style, "    background: none; \n");
    style = g_string_append (style, "    border-width: 0px; \n");
    style = g_string_append (style, "} \n");

    style = g_string_append (style, "GtkTreeView row { \n");
    if (strlen(bgcolor) > 0)       g_string_append_printf (style, "    background-color: %s; \n", bgcolor);
    if (strlen(fgcolor) > 0)       g_string_append_printf (style, "    color:            %s; \n", fgcolor);
    style = g_string_append (style, "} \n");

    style = g_string_append (style, "GtkTreeView row:selected, \n");
    style = g_string_append (style, "GtkTreeView row:active { \n");
    if (strlen(bgcolor_sel) > 0)   g_string_append_printf (style, "    background-color: %s; \n", bgcolor_sel);
    if (strlen(fgcolor_sel) > 0)   g_string_append_printf (style, "    color:            %s; \n", fgcolor_sel);
    style = g_string_append (style, "} \n");
#endif
    gchar* style_str = g_string_free (style, FALSE);

#if !GTK_CHECK_VERSION(3,0,0)
    gtk_rc_parse_string (style_str);
#else
    GtkCssProvider *css_provider = gtk_css_provider_new ();
    gtk_css_provider_load_from_data (css_provider, style_str, -1, NULL);
    GtkStyleContext *style_ctx = gtk_widget_get_style_context (widget);  // do NOT free!
    gtk_style_context_add_provider (style_ctx, GTK_STYLE_PROVIDER (css_provider), GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);
    g_object_unref (css_provider);
#endif
    g_free (style_str);
}

gboolean
tree_view_expand_rows_recursive (GtkTreeModel *model, GtkTreeView *view, GtkTreePath *parent, gint max_depth)
{
    GtkTreeIter iter;
    if (! gtk_tree_model_get_iter(model, &iter, parent))  // check if path is valid
        return FALSE;

    if (max_depth > 0 && gtk_tree_path_get_depth (parent) >= max_depth)
        return FALSE;

    // when expanding, this should come *before* going down the tree
    gtk_tree_view_expand_row (view, parent, TRUE);

    GtkTreePath *path = gtk_tree_path_copy (parent);
    gtk_tree_path_down (path);
    while (tree_view_expand_rows_recursive (model, view, path, max_depth))
        gtk_tree_path_next (path);
    gtk_tree_path_free (path);

    return TRUE;
}

gboolean
tree_view_collapse_rows_recursive (GtkTreeModel *model, GtkTreeView *view, GtkTreePath *parent, gint max_depth)
{
    GtkTreeIter iter;
    if (! gtk_tree_model_get_iter(model, &iter, parent))  // check if path is valid
        return FALSE;

    if (max_depth > 0 && gtk_tree_path_get_depth (parent) >= max_depth)
        return FALSE;

    GtkTreePath *path = gtk_tree_path_copy (parent);
    gtk_tree_path_down (path);
    while (tree_view_collapse_rows_recursive (model, view, path, max_depth))
        gtk_tree_path_next (path);
    gtk_tree_path_free (path);

    // when expanding, this should come *after* going down the tree
    gtk_tree_view_collapse_row (view, parent);

    return TRUE;
}

/* Clear nodes from tree, optionally deleting the root node */
void
tree_store_iter_clear_nodes (GtkTreeStore *store, gpointer iter, gboolean delete_root)
{
    GtkTreeIter i;
    while (gtk_tree_model_iter_children (GTK_TREE_MODEL (store), &i, iter))
    {
        if (gtk_tree_model_iter_has_child (GTK_TREE_MODEL (store), &i))
            tree_store_iter_clear_nodes (store, &i, TRUE);
        if (gtk_tree_store_iter_is_valid (GTK_TREE_STORE (store), &i))
            gtk_tree_store_remove (GTK_TREE_STORE (store), &i);
    }
    if (delete_root)
        gtk_tree_store_remove (GTK_TREE_STORE (store), iter);
}

#if GTK_CHECK_VERSION(3,6,0)
/* Get number of rows in a grid where the given column is not empty */
gint
gtk_grid_get_number_of_rows (GtkGrid *grid, gint column)
{
    int row = 0;
    while (gtk_grid_get_child_at (grid, column, row) != NULL)
    {
        row++;
    }
    return row;
}

/* Get color chooser result as hex string */
gchar *
gtk_color_chooser_get_hex (GtkColorChooser *chooser)
{
    gchar *hex;
    GdkRGBA color;

    gtk_color_chooser_get_rgba (chooser, &color);

    hex = g_strdup ("#000000");
    g_snprintf (hex, 8, "#%02x%02x%02x",
            (unsigned int)(255*color.red), (unsigned int)(255*color.green), (unsigned int)(255*color.blue));

    return hex;
}
#endif


utils.h
#include <string.h>
#include <gtk/gtk.h>
#include <gdk/gdk.h>

/* Helper macros */
#define foreach_slist_free(node,list)               \
                for (node = list, list = NULL; g_slist_free_1(list), node != NULL; list = node, node = node->next)
#define foreach_dir(filename,dir)                   \
                for ((filename) = g_dir_read_name(dir); (filename) != NULL; (filename) = g_dir_read_name(dir))
#define NZV(ptr)                                    \
                (G_LIKELY((ptr)) && G_LIKELY((ptr)[0]))
#define setptr(ptr,result)                          \
                { gpointer setptr_tmp = ptr; ptr = result; g_free(setptr_tmp); }
#define GLADE_HOOKUP_OBJECT(component,widget,name)  \
                g_object_set_data_full (G_OBJECT (component), name, gtk_widget_ref (widget), (GDestroyNotify) gtk_widget_unref)


GdkPixbuf *     utils_pixbuf_from_stock (const gchar *icon_name, gint size);
gboolean        utils_str_equal (const gchar *a, const gchar *b);
gint            utils_str_casecmp (const gchar *s1, const gchar *s2);
GSList *        utils_get_file_list_full (const gchar *path, gboolean full_path, gboolean sort, GError **error);
GSList *        utils_get_file_list (const gchar *path, guint *length, gboolean sort, GError **error);
gchar *         utils_get_utf8_from_locale(const gchar *locale_text);
void            utils_expand_home_dir (gchar *path);
gchar *         utils_tooltip_from_uri (const gchar *uri);
gchar *         utils_make_cache_path (const gchar *uri, gint imgsize, gboolean scale);
gint            utils_check_dir (const gchar *dir, mode_t mode);
void            utils_construct_style (GtkWidget *widget, const gchar *bgcolor, const gchar *fgcolor,
                const gchar *bgcolor_sel, const gchar *fgcolor_sel);

gboolean        tree_view_expand_rows_recursive (GtkTreeModel *model, GtkTreeView *view, GtkTreePath *parent, gint max_depth);
gboolean        tree_view_collapse_rows_recursive (GtkTreeModel *model, GtkTreeView *view, GtkTreePath *parent, gint max_depth);
void            tree_store_iter_clear_nodes (GtkTreeStore *store, gpointer iter, gboolean delete_root);

#if GTK_CHECK_VERSION(3,6,0)
gint            gtk_grid_get_number_of_rows (GtkGrid *grid, gint column);
gchar *         gtk_color_chooser_get_hex (GtkColorChooser *chooser);
#endif


makerelease.sh
#!/bin/bash

PACKAGENAME=deadbeef-fb

DATE=$1
FLAG=$2

if [ -z "$DATE" ]; then
    DATE=`date -u +%Y%m%d`
fi

BUILDROOT="$(pwd)"
VERSION="$(cat ${BUILDROOT}/version)"

DISTPACKAGENAME=${PACKAGENAME}-devel
INSTALLDIR=${BUILDROOT}/install
RELEASEDIR=${BUILDROOT}/release

BINTARGET=${RELEASEDIR}/binary/${PACKAGENAME}${FLAG}_${DATE}_${VERSION}.tar.gz
SRCTARGET=${RELEASEDIR}/source/${PACKAGENAME}${FLAG}_${DATE}_${VERSION}_src.tar.gz

mkdir -p ${RELEASEDIR}/binary
mkdir -p ${RELEASEDIR}/source

echo "=============================================================================="
echo "Building binary release for ${PACKAGENAME}${FLAG}-${DATE}_${VERSION} ..."

cd ${BUILDROOT}
rm -rf ${INSTALLDIR}
mkdir -p ${INSTALLDIR}/${PACKAGENAME}

./configure
make DESTDIR=${INSTALLDIR} install
libtool --finish ${INSTALLDIR}/${PACKAGENAME}

if [ -d ${INSTALLDIR} ]; then
    cd ${INSTALLDIR}
    for file in ./usr/local/lib/deadbeef/*.so.0.0.0; do
        cp -v $file ./${PACKAGENAME}/`basename $file .0.0.0` || exit $?
    done
    cp -v ${BUILDROOT}/README ./${PACKAGENAME}/ || exit $?
    cp -v ${BUILDROOT}/*install.sh ./${PACKAGENAME}/ || exit $?
    cp -v ${BUILDROOT}/*remove.sh ./${PACKAGENAME}/ || exit $?
    tar -czf ${BINTARGET} ${PACKAGENAME} || exit $?
    cd ${BUILDROOT}
fi

echo "=============================================================================="
echo "Building source release for ${PACKAGENAME}${FLAG}-${DATE}_${VERSION} ..."

cd ${BUILDROOT}
rm -f ${DISTPACKAGENAME}.tar.gz
make dist PACKAGE=${PACKAGENAME} || exit $?
mv ${DISTPACKAGENAME}.tar.gz ${SRCTARGET} || exit $?

echo "=============================================================================="
echo "Updating releases for ${PACKAGENAME}${FLAG}-${DATE}_${VERSION} ..."

cd ${RELEASEDIR}
if [ ! -d ${PACKAGENAME} ]; then
    echo "> Getting a fresh copy of the release repository ..."
    git clone -b release git@gitlab.com:zykure/${PACKAGENAME}.git || exit $?
fi
cd ${PACKAGENAME}
mkdir -p binary
mkdir -p source
cp ${BUILDROOT}/README ./README || exit $?
cp ${BINTARGET} ./binary/ || exit $?
cp ${SRCTARGET} ./source/ || exit $?
git add README || exit $?
git add binary/ || exit $?
git add source/ || exit $?
git status
echo ">>> Press CTRL+C to abort ..."
sleep 5
git commit -a -m "release ${DATE}" || exit $?
echo "> Pushing commits ..."
git pull || exit $?
git push || exit $?
cd ${BUILDROOT}


makeaur.sh
#!/bin/bash

PACKAGENAME=deadbeef-fb

DATE=$1
FLAG=$2
RELEASE=$3

if [ -z "$DATE" ]; then
    DATE=`date -u +%Y%m%d`
fi

if [ -z "$RELEASE" ]; then
    RELEASE=1
fi

BUILDROOT="$(pwd)"
VERSION="$(cat ${BUILDROOT}/version)"

RELEASEDIR=${BUILDROOT}/release
AURDIR=${BUILDROOT}/aur

SRCNAME="${PACKAGENAME}${FLAG}_${DATE}_${VERSION}_src.tar.gz"
SRCTARGET="${RELEASEDIR}/source/${SRCNAME}"

ls -l $SRCTARGET
MD5SUM=$(md5sum ${SRCTARGET} | cut -c -32)
SHA1SUM=$(sha1sum ${SRCTARGET} | cut -c -40)
SHA256SUM=$(sha256sum ${SRCTARGET} | cut -c -64)

echo "> MD5:    ${MD5SUM}"
echo "> SHA1:   ${SHA1SUM}"
echo "> SHA256: ${SHA256SUM}"

mkdir -p ${AURDIR}

function make_package
{
    AURPACKAGENAME=$1
    AURPACKAGEFLAG=$2
    AURPACKAGEDEPS=$3
    AURPACKAGECONFIG=$4

    echo "=============================================================================="
    echo "Building AUR package ${AURPACKAGENAME}${AURPACKAGEFLAG}-${DATE}_${VERSION}-${RELEASE} ..."

    cd ${BUILDROOT}
    cat PKGBUILD.in \
        | sed s/@PACKAGENAME@/${AURPACKAGENAME}/g \
        | sed s/@PACKAGEFLAG@/${AURPACKAGEFLAG}/g \
        | sed s/@PACKAGEREL@/${RELEASE}/g \
        | sed s/@PACKAGEVER@/${DATE}_${VERSION}/g \
        | sed s/@PACKAGEDEPS@/${AURPACKAGEDEPS}/g \
        | sed s/@PACKAGECONFIG@/${AURPACKAGECONFIG}/g \
        | sed s/@SOURCENAME@/${PACKAGENAME}${FLAG}_${DATE}/g \
        | sed s/@MD5SUM@/${MD5SUM}/g \
        | sed s/@SHA1SUM@/${SHA1SUM}/g \
        | sed s/@SHA256SUM@/${SHA256SUM}/g \
        > PKGBUILD
    rm -f "./${SRCNAME}"
    makepkg --source -f || exit $?
    mkdir -p ${AURDIR}/package
    mv -v ${AURPACKAGENAME}${AURPACKAGEFLAG}-${DATE}_${VERSION}-${RELEASE}.src.tar.gz ${AURDIR}/package/

    echo "=============================================================================="
    echo "Testing AUR package ${AURPACKAGENAME}${AURPACKAGEFLAG}-${DATE}_${VERSION}-${RELEASE} ..."

    rm -rf ${AURDIR}/test
    mkdir -p ${AURDIR}/test
    cd ${AURDIR}/test
    tar -xzf ${AURDIR}/package/${AURPACKAGENAME}${AURPACKAGEFLAG}-${DATE}_${VERSION}-${RELEASE}.src.tar.gz || exit $?
    ls -l
    cd ${AURPACKAGENAME}${AURPACKAGEFLAG}
    echo "> $(pwd)"
    namcap PKGBUILD || exit $?
    makepkg -f || exit $?
    namcap ${AURPACKAGENAME}${AURPACKAGEFLAG}-${DATE}_${VERSION}-${RELEASE}-any.pkg.tar.xz || exit $?
    cd ${BUILDROOT}

    echo "=============================================================================="
    echo "Updating AUR package ${AURPACKAGENAME}${AURPACKAGEFLAG}-${DATE}_${VERSION}-${RELEASE} ..."

    cd ${AURDIR}
    if [ ! -d ${AURPACKAGENAME}${AURPACKAGEFLAG} ]; then
        echo "> Getting a fresh copy of the AUR repository ..."
        git clone ssh://aurssh/${AURPACKAGENAME}${AURPACKAGEFLAG}.git || exit $?
    fi
    cd ${AURPACKAGENAME}${AURPACKAGEFLAG}
    git pull
    cp -v ${BUILDROOT}/PKGBUILD ./PKGBUILD || exit $?
    namcap PKGBUILD || exit $?
    mksrcinfo || exit $?
    git add PKGBUILD || exit $?
    git add .SRCINFO || exit $?
    git status
    echo ">>> Press CTRL+C to abort ..."
    sleep 5
    git commit -a -m "release ${DATE}-${VERSION}" || exit $?
    echo "> Pushing commits ..."
    git push || exit $?
    cd ${BUILDROOT}

}

#            name                 flag    deps   config
make_package "deadbeef-plugin-fb" ""      "gtk2" "--disable-gtk3"
make_package "deadbeef-plugin-fb" "-gtk3" "gtk3" "--disable-gtk2"
